utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+                        log.trace(() -> "Set demand to 0");
+                        break;
+                    case ON_ERROR:
+                        OnErrorQueueEntry<T> onErrorEntry = (OnErrorQueueEntry<T>) entry;
+
+                        entryTypesToFail.addAll(asList(ON_NEXT, ON_COMPLETE, ON_ERROR));
+                        log.trace(() -> "Calling onError() with " + onErrorEntry.failure, onErrorEntry.failure);
+                        subscriber.onError(onErrorEntry.failure);
+                        outstandingDemand.set(0);
+                        log.trace(() -> "Set demand to 0");
+                        break;
+                    case CANCEL:
+                        subscriber = null; // Allow subscriber to be garbage collected after cancellation.
+                        break;
+                    default:
+                        // Should never happen. Famous last words?
+                        throw new IllegalStateException("Unknown entry type: " + entry.type());
+                }
+
+                entry.resultFuture.complete(null);
+            }
+
+            eventQueue.remove();
+        }
+    }
+
+    /**
+     * Return true if we should process the provided queue entry.
+     */
+    private boolean shouldProcessQueueEntry(QueueEntry<T> entry) {
+        if (subscriber == null) {
+            // We don't have a subscriber yet.
+            return false;
+        }
+
+        if (entry == null) {
+            // The queue is empty.
+            return false;
+        }
+
+        if (entry.type() != ON_NEXT) {
+            // This event isn't an on-next event, so we don't need subscriber demand in order to process it.
+            return true;
+        }
+
+        if (entryTypesToFail.contains(ON_NEXT)) {
+            // This is an on-next call (decided above), but we're failing on-next calls. Go ahead and fail it.
+            return true;
+        }
+
+        // This is an on-next event and we're not failing on-next events, so make sure we have demand available before
+        // processing it.
+        return outstandingDemand.get() > 0;
+    }
+
+    /**
+     * Invoked from within {@link #processEventQueue()} when we can't process the queue for some reason. This is likely
+     * caused by a downstream subscriber throwing an exception from {@code onNext}, which it should never do.
+     *
+     * <p>Here we try our best to fail all of the entries in the queue, so that no callers have "stuck" futures.
+     */
+    private void panicAndDie(Throwable cause) {
+        try {
+            // Create exception here instead of in supplier to preserve a more-useful stack trace.
+            RuntimeException failure = new IllegalStateException("Encountered fatal error in publisher", cause);
+            rejectException.compareAndSet(null, () -> failure);
+            entryTypesToFail.addAll(asList(QueueEntry.Type.values()));
+            subscriber.onError(failure);
```

Comment:
This seems to be inconsistent with `#error()` and the case where request is <= in the subcription implementation where onError is not invoked immediately. Is this intentional?

Reply:
Sorry, I'm not sure what you mean. 

We call `onError` here directly instead of doing it via the queue, because `panicAndDie` is used when the queue seems to be 'broken' for some unknown reason (`doProcessQueue` threw an exception, which should never happen!). It's a last-ditch attempt to not leave any incomplete futures laying around. That's why it disables all of the entry types for processing and drains the future queue itself instead of relying on the default queue processing logic.

Reply:
Gotcha, I was wondering if we should do the same for error cases to fail faster instead of going through the queue and skipping other events

Reply:
The only reason we can do this here safely is because we're holding the "processing the queue" flag. It wouldn't be safe to call `onError` directly within `error` (we might be in the middle of calling onNext, onComplete, etc.).

Reply:
I see. In the S3 CRT publisher adapter implementation, we just clear the queue, add ON_ERROR event to the queue and then trigger processQueue so that the error event is executed almost immediately. It doesn't apply here since we are skipping events and not deleting them. This can be marked as resolved.